client/src/controllers/LinkController.ts

```diff
@@ -0,0 +1,68 @@
+import { Controller } from '@hotwired/stimulus';
+
+export class LinkController extends Controller<HTMLElement> {
+  static values = {
+    attrName: { default: 'href', type: String },
+    preserveKeys: { default: [], type: Array },
+    reflectKeys: { default: ['__all__'], type: Array },
+  };
+
+  declare attrNameValue: string;
+  declare preserveKeysValue: string[];
+  declare reflectKeysValue: string[];
+
+  get url() {
+    return new URL(
+      this.element.getAttribute(this.attrNameValue) || '',
+      window.location.href,
+    );
+  }
+
+  connect() {
+    this.setParamsFromLocation();
+  }
+
+  get reflectAll() {
+    return this.reflectKeysValue.includes('__all__');
+  }
+
+  setParamsFromURL(url: URL) {
+    // New params to build the new URL
+    const newParams = new URLSearchParams();
+    const reflectAll = this.reflectAll;
+
+    const sourceParams = url.searchParams;
+    sourceParams.forEach((value, key) => {
+      if (
+        key.startsWith('_w_') || // Wagtail internal
+        // Delete the key if we want to preserve it from the current URL, or
```

Comment:
Yeah "preserve from current URL" is an important aspect even when the param is in the current view’s query string. It’ll be good to write tests for this so we’re clearer on the correct behavior.